From f046b0bc347fe8343323404a54ef183030db99df Mon Sep 17 00:00:00 2001
From: Aseem Kumar <aseemk@google.com>
Date: Wed, 24 Jun 2026 16:10:19 -0700
Subject: [PATCH 45/49] Sanitize labels in GrantCredentialsPermissionActivity

Sanitize application labels and other displayed strings using
TextUtils.makeSafeForPresentation() to prevent UI obfuscation
attacks via newline injection.

Copied from autogenerated cl from by MendIt (go/androidmendit).

Bug: 472955222
Test: m framework-minus-apex
Flag: EXEMPT BUGFIX
(cherry picked from commit d143444ef510fb4520245a8cd69215278928fec5)
Cherrypick-From: https://googleplex-android-review.googlesource.com/q/commit:1fe77219000258a2dfccbf5eea25d1f1e9f334ea
Merged-In: Ief6f22746200360e97b3ccbc0b5abca8867139bb
Change-Id: Ief6f22746200360e97b3ccbc0b5abca8867139bb
---
 .../GrantCredentialsPermissionActivity.java   | 20 +++++++++++++++----
 1 file changed, 16 insertions(+), 4 deletions(-)

diff --git a/core/java/android/accounts/GrantCredentialsPermissionActivity.java b/core/java/android/accounts/GrantCredentialsPermissionActivity.java
index d67bff2c02b2..0925e6014e53 100644
--- a/core/java/android/accounts/GrantCredentialsPermissionActivity.java
+++ b/core/java/android/accounts/GrantCredentialsPermissionActivity.java
@@ -18,6 +18,7 @@ package android.accounts;
 import android.app.Activity;
 import android.content.Context;
 import android.content.Intent;
+import android.content.pm.PackageItemInfo;
 import android.content.pm.PackageManager;
 import android.content.res.Resources;
 import android.graphics.Insets;
@@ -126,7 +127,7 @@ public class GrantCredentialsPermissionActivity extends Activity implements View
                         runOnUiThread(new Runnable() {
                             public void run() {
                                 if (!isFinishing()) {
-                                    authTokenTypeView.setText(authTokenLabel);
+                                    authTokenTypeView.setText(makeSafe(authTokenLabel));
                                     authTokenTypeView.setVisibility(View.VISIBLE);
                                 }
                             }
@@ -156,11 +157,22 @@ public class GrantCredentialsPermissionActivity extends Activity implements View
             } catch (PackageManager.NameNotFoundException e) {
                 packageLabel = pkg;
             }
-            packagesListView.addView(newPackageView(packageLabel));
+            packagesListView.addView(newPackageView(makeSafe(packageLabel)));
         }
 
-        ((TextView) findViewById(R.id.account_name)).setText(mAccount.name);
-        ((TextView) findViewById(R.id.account_type)).setText(accountTypeLabel);
+        ((TextView) findViewById(R.id.account_name)).setText(makeSafe(mAccount.name));
+        ((TextView) findViewById(R.id.account_type)).setText(makeSafe(accountTypeLabel));
+    }
+
+    private String makeSafe(String label) {
+        if (TextUtils.isEmpty(label)) {
+            return label;
+        }
+        return TextUtils.makeSafeForPresentation(label,
+                PackageItemInfo.MAX_SAFE_LABEL_LENGTH,
+                PackageItemInfo.DEFAULT_MAX_LABEL_SIZE_PX,
+                TextUtils.SAFE_STRING_FLAG_TRIM | TextUtils.SAFE_STRING_FLAG_FIRST_LINE)
+                .toString();
     }
 
     private String getAccountLabel(Account account) {
-- 
2.55.0

